### PR TITLE
feat: better handling of OpenAPI tools

### PR DESCRIPTION
**Pull request by g-linville**
This changes how we handle OpenAPI tools. Rather than generating an individual tool for each operation in the OpenAPI file, we always generate three tools: `list`, `getSchema`, and `run`. These basically act as an interface around the OpenAPI file. The LLM calls `list` to find out about all of the available operations. It then calls `getSchema` to get the JSONSchema for the args for a particular operation. It can then call that operation with the `run` tool by providing the name of the operation and the JSON string for the args.

Right now this is hidden behind the `GPTSCRIPT_OPENAPI_REVAMP` environment variable.

### CODE REVIEW

File: pkg/openapi/list.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+				err   error
+			)
+			if filter != "" && filter != NoFilter {
+				if strings.Contains(filter, "*") {
+					var filters []string
+					if strings.Contains(filter, "|") {
+						filters = strings.Split(filter, "|")
+					} else {
+						filters = []string{filter}
+					}
```
Comment (iwilltry42) [inline pkg/openapi/list.go:36]:
```suggestion
					filters := strings.Split(filter, "|")
```

If `|` is not in there, Split will just return `[]string{filter}` anyway

File: pkg/loader/loader.go [hunk 2]
```diff
@@ -157,7 +157,7 @@ func loadOpenAPI(prg *types.Program, data []byte) *openapi3.T {
 		prg.OpenAPICache = map[string]any{}
 	}
 
-	switch isOpenAPI(data) {
```
Comment (iwilltry42) [inline pkg/loader/loader.go:160]: It should be possible to use `openapi.LoadFromBytes` (if exported) here and just add the cache part after it, right? Not sure if it's worth it for the transition period :man_shrugging:

File: pkg/openapi/run.go [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+	}
+
+	// Make the request
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return "", false, fmt.Errorf("failed to make request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	result, err := io.ReadAll(resp.Body)
```
Comment (iwilltry42) [inline pkg/openapi/run.go:150]: To be on the safe side, this could use a nil check to ensure that the Body exists

**Review by iwilltry42 — COMMENTED** [review]
This PR is a monster O.o 
Good Job! 
Left some comments...

File: pkg/openapi/list.go (hunk 1 above)
Comment (g-linville) [reply pkg/openapi/list.go]: I feel silly for writing it this way in the first place, lol

**Review by iwilltry42 — APPROVED** [review]

File: pkg/openapi/run.go [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+	}
+
+	// Make the request
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return "", false, fmt.Errorf("failed to make request: %w", err)
+	}
+
+	var result []byte
+	if resp.Body != nil {
```
Comment (thedadams) [inline pkg/openapi/run.go:152]: `resp.Body` is always non-nil if the `err` above is nil.

File: pkg/openapi/getschema.go [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+	SecurityInfos   [][]SecurityInfo `json:"securityInfos"`
+	QueryParams     []Parameter      `json:"queryParameters"`
+	PathParams      []Parameter      `json:"pathParameters"`
+	HeaderParams    []Parameter      `json:"headerParameters"`
+	CookieParams    []Parameter      `json:"cookieParameters"`
+}
+
+var (
+	SupportedMIMETypes     = []string{"application/json", "application/x-www-form-urlencoded", "multipart/form-data"}
+	SupportedSecurityTypes = []string{"apiKey", "http"}
```
Comment (thedadams) [inline pkg/openapi/getschema.go:32]: nit: my feeling is that it is always better to have non-const package level variables by unexported and use a getter.

**Review by thedadams — COMMENTED** [review]
Mostly just nits.

File: pkg/engine/openapi.go [hunk 6]
```diff
@@ -8,90 +8,159 @@ import (
...
+			return nil, fmt.Errorf("failed to list operations: %w", err)
+		}
+
+		opListJSON, err := json.MarshalIndent(opList, "", "    ")
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal operation list: %w", err)
+		}
+
+		res = &Return{
+			Result: ptr(string(opListJSON)),
```
[the remark this answers is not in the document]
Comment (g-linville) [reply pkg/engine/openapi.go:57]: same

File: pkg/openapi/getschema.go (hunk 5 above)
Comment (g-linville) [reply pkg/openapi/getschema.go]: Done

File: pkg/openapi/run.go (hunk 4 above)
Comment (g-linville) [reply pkg/openapi/run.go]: Fixed

File: pkg/openapi/run.go [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+
+	// Handle request body
+	if opInfo.BodyContentMIME != "" {
+		res := gjson.Get(args, "requestBodyContent")
+		var body bytes.Buffer
+		switch opInfo.BodyContentMIME {
+		case "application/json":
+			var reqBody interface{}
+
+			reqBody = struct{}{}
```
[the remark this answers is not in the document]
Comment (g-linville) [reply pkg/openapi/run.go]: Done

File: pkg/openapi/run.go [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+
+	if args == "" {
+		args = "{}"
+	}
+	schemaJSON, opInfo, found, err := GetSchema(operationID, defaultHost, t)
+	if err != nil {
+		return "", false, err
+	} else if !found {
+		return "", false, nil
+	}
```
[the remark this answers is not in the document]
Comment (g-linville) [reply pkg/openapi/run.go]: Done

File: pkg/openapi/run.go [hunk 9]
```diff
@@ -0,0 +1,459 @@
...
+	"github.com/gptscript-ai/gptscript/pkg/env"
+	"github.com/tidwall/gjson"
+	"github.com/xeipuuv/gojsonschema"
+	"golang.org/x/exp/maps"
+)
+
+const RunTool = "run"
+
+func Run(operationID, defaultHost, args string, t *openapi3.T, envs []string) (string, bool, error) {
+	envMap := map[string]string{}
```
[the remark this answers is not in the document]
Comment (g-linville) [reply pkg/openapi/run.go]: Done

File: pkg/openapi/load.go [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+func LoadFromBytes(content []byte) (*openapi3.T, error) {
+	var (
+		openAPIDocument *openapi3.T
+		err             error
+	)
+
+	switch IsOpenAPI(content) {
+	case 2:
+		// Convert OpenAPI v2 to v3
+		jsonContent := content
```
[the remark this answers is not in the document]
Comment (g-linville) [reply pkg/openapi/load.go]: Nope not sure why it was this way. Fixed

File: pkg/engine/openapi.go [hunk 11]
```diff
@@ -8,90 +8,159 @@ import (
...
 	"github.com/tidwall/gjson"
-	"golang.org/x/exp/maps"
 )
 
-var (
-	SupportedMIMETypes     = []string{"application/json", "text/plain", "multipart/form-data"}
-	SupportedSecurityTypes = []string{"apiKey", "http"}
-)
+func (e *Engine) runOpenAPIRevamp(tool types.Tool, input string) (*Return, error) {
+	envMap := map[string]string{}
```
[the remark this answers is not in the document]
Comment (g-linville) [reply pkg/engine/openapi.go]: Done

**Review by thedadams — APPROVED** [review]

File: pkg/openapi/load.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (iwilltry42) [reply pkg/openapi/load.go]: You asked that before @thedadams :wink: 
https://github.com/gptscript-ai/gptscript/pull/345#discussion_r1608829599

**Pull request merged**